File: Java/src/main/java/com/DSA/Java/Algorithms/Searching/BinarySearch.java
package com.DSA.Java.Algorithms.Searching;

public class BinarySearch {
  private int[] arr;

  public BinarySearch(int[] arr) {
    this.arr = arr;
  }

  public void search(int start, int end, int value) {
    int mid = (start + end) / 2;
    if (start > end) {
      System.out.println("The value " + value + " was not found in the array");
      return ;
    }

    if (arr[mid] == value) {
      System.out.println("The value " + value + " is found on the index " + mid + " in the array.");
      return ;
    } else if (value < arr[mid]) {
      search(start, mid - 1, value);
    } else if (value > arr[mid]) {
      search(mid + 1, end, value);
    }
  }
}

File: C++/Algorithms/src/Searching/BinarySearch.cpp
#include "../../include/Searching/BinarySearch.h"

void search_element_helper(int start, int end, int value, std::vector<int> &arr) {
    int midpoint = (start + end) / 2;
    if (start > end) {
        std::cout << "The value was not found" << std::endl;
        return;
    }

    if (arr[midpoint] == value) {
        std::cout << "The value " << value << " was found on index " << midpoint << std::endl;
        return;
    } else if (arr[midpoint] > value) {
        search_element_helper(start, midpoint - 1, value, arr);
    } else if (arr[midpoint] < value) {
        search_element_helper(midpoint + 1, end, value, arr);
    }
}

void BinarySearch::search_element(int value) {
    search_element_helper(0, arr.size(), value, arr);
}
